test/controllers/auth_controller_test.rb

```diff
@@ -2,32 +2,108 @@
 
 require 'test_helper'
 
-class HealthCardsControllerTest < ActionDispatch::IntegrationTest
-  test 'authorize without client_id should return 401' do
+class AuthControllerTest < ActionDispatch::IntegrationTest
+  test 'authorize without client_id should return 400 invalid_request' do
     get(auth_code_path)
-    assert_response :unauthorized
+    assert_response :bad_request
+    assert_equal('{"error":"invalid_request"}', @response.body)
   end
 
-  test 'authorize with incorrect client_id should return 401' do
-    get(auth_code_path, params: { client_id: 'id' })
-    assert_response :unauthorized
+  test 'authorize with incorrect client_id should return 400 invalid_request' do
+    get(auth_code_path, params: { client_id: 'bad_id' })
+    assert_response :bad_request
+    assert_equal('{"error":"invalid_request"}', @response.body)
   end
 
-  test 'authorize with correct client_id redirects with auth_code' do
+  test 'authorize with correct parameters should redirect with auth_code' do
     redirect_uri = 'http://example.com'
     state = 'test_state'
     get(auth_code_path,
-        params: { client_id: Rails.application.config.client_id, redirect_uri: redirect_uri, state: state })
+        params: { client_id: Rails.application.config.client_id,
+                  code: Rails.application.config.auth_code,
+                  redirect_uri: redirect_uri,
+                  state: state })
     assert_redirected_to "#{redirect_uri}?code=#{Rails.application.config.auth_code}&state=#{state}"
   end
 
-  test 'token with incorrect code' do
-    post(auth_token_path, params: { code: 'code' })
-    assert_response :unauthorized
+  test 'token with no parameters should return 400 invalid_request' do
+    post(auth_token_path)
+    assert_response :bad_request
+    assert_equal('{"error":"invalid_request"}', @response.body)
   end
 
-  test 'token with correct code' do
-    post(auth_token_path, params: { code: Rails.application.config.auth_code })
+  test 'token with missing client_id parameter should return 400 invalid_request' do
+    post(auth_token_path, params: { code: Rails.application.config.client_id })
```

Comment:
With these params, it's difficult to reason about exactly what is causing the errors. I think it would be better to have a set of good params, and then modify them in each test so that the parameters are good except for the specific ones that need to be bad for a particular test. Like:
```
post(auth_token_path, params: good_params.merge(code: 'BAD_CODE'))
```

Reply:
I will also add this to next commit.